src/NewlineLexers.jl

```diff
@@ -123,13 +130,36 @@ end
 # These must be respected in the `_find_newlines_kernel!` and `_find_newlines_generic!`
 # All other cases are unambiguous, i.e. we can tell if we are inside a string or not,
 # and if we are, we can tell if we are on an escape or not.
+"""
+    Lexer{E,OQ,CQ,NL,IO_t}
+
+    Lexer(io, escapechar, openquotechar, closequotechar, newline) -> Lexer{E,OQ,CQ,NL,IO_t}
+    Lexer(io, nothing, newline) -> Lexer{Nothing,Nothing,Nothing,NL,IO_t}
+
+A steteful lexer type for newline detection. Use with the `find_newlines!` function.
+The type parameters are:
+
+- `E`: the escape character
+- `OQ`: the open quote character
+- `CQ`: the close quote character
+- `NL`: the newline character
+- `IO_t`: the type of the IO object, e.g. `IOBuffer` or `IOStream`
+
+Only single-byte characters are supported.
+When `E`, `OQ`, and `CQ` are not `Nothing`, the lexer will find all newlines in the input,
```

Comment:
what's the dependence between `E`, `OQ`, and `CQ`? is it that they must _all_ be `Nothing` or _all_ be non-`Nothing`?